File: windows-native/src/Webyar.Core/Localization/Strings.cs
using System.Globalization;
using System.Text.Json;
using System.Text.RegularExpressions;

namespace Webyar.Core.Localization;

public enum Language
{
    Fa,
    En,
    Tr,
}

/// <summary>
/// The app's copy in Persian, English and Turkish — the same sentences the iOS
/// app and the web console use (see scripts/export-strings.mjs). `{name}`
/// placeholders are filled in, numbers in the language's own digits.
/// </summary>
public sealed partial class Strings
{
    private readonly Dictionary<Language, Dictionary<string, string>> _table;

    public Strings(Language language)
    {
        Language = language;
        _table = Shared.Value;
    }

    private static readonly Lazy<Dictionary<Language, Dictionary<string, string>>> Shared = new(Load);

    public Language Language { get; }

    public bool IsRightToLeft => Language == Language.Fa;

    public CultureInfo Culture => CultureOf(Language);

    public static CultureInfo CultureOf(Language language) => language switch
    {
        Language.Fa => CultureInfo.GetCultureInfo("fa-IR"),
        Language.Tr => CultureInfo.GetCultureInfo("tr-TR"),
        _ => CultureInfo.GetCultureInfo("en-US"),
    };

    public static string Code(Language language) => language switch
    {
        Language.Fa => "fa",
        Language.Tr => "tr",
        _ => "en",
    };

    public static Language? Parse(string? code) => code?.Trim().ToLowerInvariant() switch
    {
        "fa" or "fa-ir" => Language.Fa,
        "tr" or "tr-tr" => Language.Tr,
        "en" or "en-us" or "en-gb" => Language.En,
        _ => null,
    };

    /// <summary>The first launch follows Windows' language when it is one of ours, Persian otherwise.</summary>
    public static Language FromSystem(CultureInfo culture) =>
        Parse(culture.TwoLetterISOLanguageName) ?? Language.Fa;

    public string this[string key] => Get(key);

    public string Get(string key, IReadOnlyDictionary<string, object>? args = null)
    {
        var raw = Lookup(Language, key) ?? Lookup(Language.En, key) ?? key;
        if (args is null || args.Count == 0) return raw;
        return Placeholder().Replace(raw, m =>
        {
            if (!args.TryGetValue(m.Groups[1].Value, out var value)) return m.Value;
            return value switch
            {
                int i => Digits.Localize(i.ToString(CultureInfo.InvariantCulture), Language),
                long l => Digits.Localize(l.ToString(CultureInfo.InvariantCulture), Language),
                double d => Digits.Localize(d.ToString(CultureInfo.InvariantCulture), Language),
                _ => value?.ToString() ?? string.Empty,
            };
        });
    }

    public string Get(string key, string name, object value) =>
        Get(key, new Dictionary<string, object> { [name] = value });

    public bool Has(string key) => Lookup(Language.En, key) is not null;

    private string? Lookup(Language language, string key) =>
        _table.TryGetValue(language, out var dict) && dict.TryGetValue(key, out var s) && s.Length > 0 ? s : null;

    [GeneratedRegex(@"\{(\w+)\}")]
    private static partial Regex Placeholder();

    private static Dictionary<Language, Dictionary<string, string>> Load()
    {
        using var stream = typeof(Strings).Assembly.GetManifestResourceStream("Webyar.Core.strings.json")
            ?? throw new InvalidOperationException("strings.json is not embedded");
        var raw = JsonSerializer.Deserialize<Dictionary<string, Dictionary<string, string>>>(stream)
            ?? throw new InvalidOperationException("strings.json is empty");
        var table = new Dictionary<Language, Dictionary<string, string>>();
        foreach (var (code, dict) in raw)
        {
            if (Parse(code) is { } language) table[language] = dict;
        }
        return table;
    }
}

/// <summary>Western digits to Persian ones, the way every Persian UI in the product writes numbers.</summary>
public static class Digits
{
    private const string Persian = "۰۱۲۳۴۵۶۷۸۹";

    public static string Localize(string text, Language language)
    {
        if (language != Language.Fa) return text;
        return string.Create(text.Length, text, (span, source) =>
        {
            for (var i = 0; i < source.Length; i++)
            {
                var c = source[i];
                span[i] = c is >= '0' and <= '9' ? Persian[c - '0'] : c;
            }
        });
    }
}

File: macos/Webyar/Core/Localization/Strings.swift
import Foundation

enum Language: String, CaseIterable, Sendable, Identifiable {
    case fa, en, tr

    var id: String { rawValue }

    var code: String { rawValue }

    var isRightToLeft: Bool { self == .fa }

    /// Its own name, as the language picker lists it.
    var nativeName: String {
        switch self {
        case .fa: return "فارسی"
        case .en: return "English"
        case .tr: return "Türkçe"
        }
    }

    var locale: Locale {
        switch self {
        case .fa: return Locale(identifier: "fa_IR")
        case .tr: return Locale(identifier: "tr_TR")
        case .en: return Locale(identifier: "en_US")
        }
    }

    static func parse(_ code: String?) -> Language? {
        switch code?.trimmingCharacters(in: .whitespaces).lowercased() {
        case "fa", "fa-ir", "fa_ir": return .fa
        case "tr", "tr-tr", "tr_tr": return .tr
        case "en", "en-us", "en-gb", "en_us", "en_gb": return .en
        default: return nil
        }
    }

    /// The first launch follows the Mac's language when it is one of ours, Persian otherwise.
    static var system: Language {
        for id in Locale.preferredLanguages {
            if let l = parse(String(id.prefix(2))) { return l }
        }
        return .fa
    }
}

/// The app's copy in Persian, English and Turkish — the same sentences the
/// Windows, iOS and web apps use. The table is the Windows app's own
/// strings.json (windows-native/src/Webyar.Core/Localization), bundled as is,
/// with the few Mac-specific lines layered on top (Resources/mac-strings.json).
/// `{name}` placeholders are filled in, numbers in the language's own digits.
struct Strings: Sendable {
    let language: Language
    private let table: [String: [String: String]]

    init(_ language: Language, table: [String: [String: String]]? = nil) {
        self.language = language
        self.table = table ?? Self.shared
    }

    var isRightToLeft: Bool { language.isRightToLeft }

    subscript(_ key: String) -> String { self.get(key) }

    func get(_ key: String, _ args: [String: Any] = [:]) -> String {
        let raw = lookup(language, key) ?? lookup(.en, key) ?? key
        guard !args.isEmpty else { return raw }
        var out = ""
        var i = raw.startIndex
        while i < raw.endIndex {
            if raw[i] == "{", let close = raw[i...].firstIndex(of: "}") {
                let name = String(raw[raw.index(after: i)..<close])
                if !name.isEmpty, name.allSatisfy({ $0.isLetter || $0.isNumber || $0 == "_" }), let value = args[name] {
                    out += format(value)
                    i = raw.index(after: close)
                    continue
                }
            }
            out.append(raw[i])
            i = raw.index(after: i)
        }
        return out
    }

    func get(_ key: String, _ name: String, _ value: Any) -> String { get(key, [name: value]) }

    func has(_ key: String) -> Bool { lookup(.en, key) != nil }

    /// A number in this language's digits.
    func number(_ n: Int) -> String { Digits.localize(String(n), language) }

    private func format(_ value: Any) -> String {
        switch value {
        case let i as Int: return Digits.localize(String(i), language)
        case let i as Int64: return Digits.localize(String(i), language)
        case let d as Double: return Digits.localize(d == d.rounded() ? String(Int64(d)) : String(d), language)
        default: return "\(value)"
        }
    }

    private func lookup(_ language: Language, _ key: String) -> String? {
        guard let s = table[language.code]?[key], !s.isEmpty else { return nil }
        return s
    }

    static let shared: [String: [String: String]] = load()

    private static func load() -> [String: [String: String]] {
        var table: [String: [String: String]] = [:]
        // The shared table first, then every "*-strings.json" the Mac app adds on top.
        let bundle = Bundle.main
        var urls: [URL] = []
        if let base = bundle.url(forResource: "strings", withExtension: "json") { urls.append(base) }
        urls += (bundle.urls(forResourcesWithExtension: "json", subdirectory: nil) ?? [])
            .filter { $0.deletingPathExtension().lastPathComponent.hasSuffix("-strings") }
            .sorted { $0.lastPathComponent < $1.lastPathComponent }
        for url in urls {
            guard let data = try? Data(contentsOf: url),
                  let raw = try? JSONSerialization.jsonObject(with: data) as? [String: [String: String]] else { continue }
            for (code, dict) in raw { table[code, default: [:]].merge(dict) { _, new in new } }
        }
        return table
    }
}


/// Western digits to Persian ones, the way every Persian UI in the product writes numbers.
enum Digits {
    private static let persian: [Character] = ["۰", "۱", "۲", "۳", "۴", "۵", "۶", "۷", "۸", "۹"]

    static func localize(_ text: String, _ language: Language) -> String {
        guard language == .fa else { return text }
        return String(text.map { c in
            if let d = c.wholeNumberValue, c.isASCII { return persian[d] }
            return c
        })
    }
}
